Various slider-related difficulty changes proposal

Givikap120 opened the pull request:
List of the changes:
- Sliderless velocity is now calculated correctly, using JumpDistance instead of LazyJumpDistance
- Stream spacing now calculated correctly, using JumpDistance instead of MinimalJumpDistance+TravelDistance
- Penalty for missing sliderends now calculated correctly
- Aim now have 1.08 multiplier for sliderjumps
- Speed now have 1.25 multiplier for sliderstreams
- HD now awards slideraim 2 times less (oshama scramble EZHD fix)

What to expect:
- Oshama Scramble EZHD is nerfed
- Sliderjumps maps are buffed (https://osu.ppy.sh/beatmapsets/656918#osu/1391631 is 6.72*)
- Sliderstream maps are buffed (https://osu.ppy.sh/beatmapsets/881996#osu/1844019 is 7.17*)

This rework doesn't fix the sliders completely, and it doesn't touch complex slidershapes problem. It's main point is to fix the some flaws of the current slider system, until better system can be made.

Givikap120 committed:
Slider aim rework

Givikap120 committed:
adjusted multipliers

to keep the same scale

Givikap120 commented:
![image](https://github.com/ppy/osu/assets/89256026/6dce51bb-b1b7-4aa9-976f-85c372eb64bf)

peppy commented:
- Please change the title to something more descriptive.
- Please link discussions where these changes were decided.

Givikap120 commented:
> * Please change the title to something more descriptive.
> * Please link discussions where these changes were decided.

This is not "decided", this is rework proposal (rework proposals are not in the osu PRs anymore?)

Givikap120 committed:
Update AimEvaluator.cs

added angle and velocity bonuses to sliderjump
fixed alternating bonus

Givikap120 committed:
Update AimEvaluator.cs

Xexxar commented:
While I do believe that sliders are hugely problematic in their current state, I personally don't have much feedback to give on this PR as I believe a more fundamental rewrite of sliders is needed to properly assess their difficulty anyway, which is an ongoing effort of mine. This likely does improve things from a values standpoint, but it seems like something that would only be a stop gap solution until something more elaborate can be created. With that in mind, I personally don't see the need to push this within the context of what else is being worked on.

Givikap120 commented:
> While I do believe that sliders are hugely problematic in their current state, I personally don't have much feedback to give on this PR as I believe a more fundamental rewrite of sliders is needed to properly assess their difficulty anyway, which is an ongoing effort of mine. This likely does improve things from a values standpoint, but it seems like something that would only be a stop gap solution until something more elaborate can be created. With that in mind, I personally don't see the need to push this within the context of what else is being worked on.

this rework is a free value improve if we won't have new slider aim algo in next deploy

smoogipoo commented:
!diffcalc

Givikap120 committed:
Merge branch 'ppy:master' into slider_jumps

Givikap120 committed:
minor changes to sliderstreams

Givikap120 committed:
Fixed sliderstreams

it used to depend on velocity, now it depends on spacing

smoogipoo commented:
!diffcalc

Givikap120 committed:
fixed high CS bug

now it's applied correctly

Givikap120 committed:
Added bandaid to fix one sliderjump cases

Most notable is on this map (was 8.2* in previous version)
https://osu.ppy.sh/beatmapsets/2005921#osu/4172054

Givikap120 committed:
Merge branch 'ppy:master' into slider_jumps

Givikap120 committed:
Fixed the bug

Givikap120 committed:
now sliderend penalty is applied correctly

Givikap120 commented:
> !diffcalc
> 
> Good time to test out some updates I've made to the sheet generator :)

bro's using my PR as a test 💀
thanks anyway

Givikap120 commented:
Will be revived to merge this with reading pp

Givikap120 committed:
reverted slideraim changes

Givikap120 committed:
Update Slider.cs

Givikap120 commented:
!diffcalc

smoogipoo commented:
!diffcalc

smoogipoo commented:
!diffcalc
SCORE_PROCESSOR_A=https://github.com/ppy/osu-queue-score-statistics/pull/263
SCORE_PROCESSOR_B=https://github.com/ppy/osu-queue-score-statistics/pull/263

Givikap120 committed:
fixed sliderstreams having 0 spacing

and changed speed sliderstream bonus

Givikap120 committed:
fixed HD bonus for slideraim

Givikap120 committed:
Update AimEvaluator.cs

Givikap120 committed:
Update SpeedEvaluator.cs

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -34,19 +36,24 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
-            double currVelocity = osuCurrObj.LazyJumpDistance / osuCurrObj.StrainTime;
+            double currVelocity = osuCurrObj.JumpDistance / osuCurrObj.StrainTime;
+
+            // Bandaid to prevent Lost Umbrella from gaining
```
I don't really like to imagine this as a "Lost Umbrella bandaid" but rather a change to stop over-rating patterns that appear in this beatmap (and other similar offenders)...

Givikap120 replied:
In general - this multiplier exists to prevent slideraim from getting pp for sliderjump
So sliderjump bonus will be only for circle aim

tsunyoku replied:
Yes, I'm implying you should rephrase the comment/remove it

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +118,50 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+                    // Take just slider heads into account because we're computing sliderjumps, not slideraim
+                    sliderJumpBonus = slider_jump_multiplier * aimStrain * sliderJumpsAdjustingMultiplier;
+
+                    // Reward more if sliders and circles are alternating (actually it's still lower than several sliders in a row)
```
What is the rationale behind this? If we're alternating between sliders and circles then you're saying it's more difficult to aim than being slider jumps only.. but why?

Givikap120 replied:
in general - the meaning of buffing slider jumps is accounting for the fact that slider jumps are more difficult to process/read in comparison to normal jumps
but the cases where there's some notes in the middle of slider jump section or vice versa - it's not becoming any easier to read than just all sliders

Givikap120 replied:
you can create jumps test map with three cases:
- only circles
- alternating
- only sliders
in my experience last two have similar difficulty, but maybe i'm just skill-issued

Givikap120 replied:
I just tested the jumps here:
With alternating bonus:
- normal - 3.38*
- alternating - 3.48*
- sliders - 3.58

Without:
- normal - 3.38*
- alternating - 3.45*
- sliders - 3.58

Givikap120 replied:
maybe i actually should revert alternating multiplier to 1.0x

Givikap120 replied:
even with 2.0x it's still only 3.55*

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -86,7 +86,20 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
         private double computeAimValue(ScoreInfo score, OsuDifficultyAttributes attributes)
         {
-            double aimValue = Math.Pow(5.0 * Math.Max(1.0, attributes.AimDifficulty / 0.0675) - 4.0, 3.0) / 100000.0;
+            double aimDifficulty = attributes.AimDifficulty;
+
+            // We assume 25% of sliders in a map are difficult since there's no way to tell from the performance calculator.
+            double estimateDifficultSliders = attributes.SliderCount * 0.25;
+
+            if (attributes.SliderCount > 0)
+            {
+                // Meh and Miss are more severe on sliders than normal Ok, so punish them 2 times more
+                double estimateSliderEndsDropped = Math.Clamp(Math.Min(countOk + countMeh * 2 + countMiss * 2, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
+                double sliderNerfFactor = (1 - attributes.SliderFactor) * (1 - estimateSliderEndsDropped / estimateDifficultSliders) + attributes.SliderFactor;
+                aimDifficulty *= sliderNerfFactor;
+            }
+
+            double aimValue = Math.Pow(5.0 * Math.Max(1.0, aimDifficulty / 0.0675) - 4.0, 3.0) / 100000.0;
```
I'm not really sure about this entire section. Something seems weird about adjusting the aim difficulty attributes rather than adjusting the aim value itself. It ruins the model of base values always being the same for a star rating but I might just be picky, open to other opinions.

Givikap120 replied:
moving sliderend punishments to the difficulty attribute makes perfect sense from the fact what `SliderFactor` is

apollo-dw replied:
I would prefer it to be cubic here.

Givikap120 replied:
> I would prefer it to be cubic here.
okay, i will change it then

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +118,50 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+            aimStrain += Math.Max(acuteAngleBonus * acute_angle_multiplier, wideAngleBonus * wide_angle_multiplier + velocityChangeBonus * velocity_change_multiplier);
+
+            double sliderJumpBonus = 0;
             {
```
Don't *really* want to pick at formatting but can you get rid of the useless braces around this block...

Givikap120 replied:
Yes, this exists from the old versions where it haven't compiled without the braces

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -34,19 +36,24 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
                 double movementVelocity = osuCurrObj.MinimumJumpDistance / osuCurrObj.MinimumJumpTime; // calculate the movement velocity from slider end to current object
 
-                currVelocity = Math.Max(currVelocity, movementVelocity + travelVelocity); // take the larger total combined velocity.
+                double newVelocity = Math.Max(currVelocity, movementVelocity + travelVelocity); // take the larger total combined velocity.
+                if (currVelocity > 0) sliderJumpsAdjustingMultiplier = currVelocity / newVelocity;
+                currVelocity = newVelocity;
```
```suggestion
                double newVelocity = Math.Max(currVelocity, movementVelocity + travelVelocity); // take the larger total combined velocity.

                if (currVelocity > 0)
                    sliderJumpsAdjustingMultiplier = currVelocity / newVelocity;

                currVelocity = newVelocity;
```

This really confused me at first read, please make it a little clearer like so..

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +118,50 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+                        sliderJumpBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuLastObj.StrainTime), 2);
+
+                    // Punish too short sliders to prevent cheesing (cheesing is still possible, but it's very rare)
+                    double sliderLength = slider.Velocity * slider.SpanDuration;
```
There's an inconsistency here, in the speed evaluator you are taking `sliderLength` as `2 * slider.Velocity * slider.SpanDuration`. I assume this isn't intended, and if it is then it's weird.

Givikap120 replied:
i just want to punish short sliders in speed 2 times less than in aim
but i will fix the fact that it's looking confusing

Givikap120 committed:
refactoring

Givikap120 committed:
Update OsuPerformanceCalculator.cs

Givikap120 committed:
Update AimEvaluator.cs

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +121,48 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+                if (osuCurrObj.StrainTime > osuLastObj.StrainTime)
+                    sliderJumpBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuLastObj.StrainTime), 2);
+
+                // Punish too short sliders to prevent cheesing (cheesing is still possible, but it's very rare)
+                double sliderLength = slider.Velocity * slider.SpanDuration;
+                if (sliderLength < slider.Radius)
+                    sliderJumpBonus *= sliderLength / slider.Radius;
             }
```
rename this variable to `sliderDuration` as length can ambiguously be understood to mean distance here.

i would be expecting something to do with follow circle radius here (2.4x). did you just not find that to be necessary? your comment does state that cheesing still occurs with this code, so something to consider

Givikap120 replied:
sliderjumps is not connected with sliderbody difficulty in any way, so there's no sense in using follow circle radius
cheesing can occur cuz you can hide long sliderbody under the circle, so the slider will appear like a circle
but i haven't seen it in ranked maps so it's good

Givikap120 replied:
also, it's indeed length, not a duration
do you misunderstanding something?
velocity * time = distance
velocity * duration = length

apollo-dw replied:
> sliderjumps is not connected with sliderbody difficulty in any way, so there's no sense in using follow circle radius

but slider leniency will allow short enough sliders to effectively allow the movement as if it were circles. mappers have used this technique for over a decade at this point

> also, it's indeed length, not a duration
yeah I think I just get tripped up by people using the term "length of time" in sentences. duration is just less ambiguous IMO

Givikap120 replied:
> but slider leniency will allow short enough sliders to effectively allow the movement as if it were circles. mappers have used this technique for over a decade at this point

this is literally the point of sliderjumps
even tho technically they allow the same movement - they're still harder because they're harder to read

> yeah I think I just get tripped up by people using the term "length of time" in sentences. duration is just less ambiguous IMO

how duration is less ambiguous if duration literally means time, when this is clearly length in pixels

apollo-dw replied:
... i think i confused myself on the duration and length thing again. i just read myself and i am not making any sense, disregard

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +121,48 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
             // Add in additional slider velocity bonus.
-            if (withSliderTravelDistance)
-                aimStrain += sliderBonus * slider_multiplier;
+            double sliderBonus = 0;
+
+            if (withSliderTravelDistance && osuLastObj.BaseObject is Slider)
+            {
+                // Reward sliders based on velocity.
+                sliderBonus = osuLastObj.TravelDistance / osuLastObj.TravelTime;
+            }
```
maybe rename this to `sliderBodyBonus` to differentiate further from `sliderJumpBonus`. and then rename `slider_aim_multiplier` to `slider_body_multiplier`

Givikap120 replied:
ok

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +121,48 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+                    sliderJumpBonus += alternatingBonus;
+                }
+
+                // If slider was slower than notes before - punish it
+                if (osuCurrObj.StrainTime > osuLastObj.StrainTime)
+                    sliderJumpBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuLastObj.StrainTime), 2);
```
i'm confused by the comment and formula here. the comment talks about notes before the slider, but the slider is `osuLastObj`. so wouldn't the notes before be `osuLastLastObj`? 

the comment could be less ambiguous too since when i read "slider was slower", i am interpreting this as the slider's velocity. but the code seems to be regarding a slowing down in rhythm

Givikap120 replied:
this line exist to not buff the cases like "jumps jumps jumps, 1/2 slider, jumps jumps jumps"
now I'm think about it - it's still have 0.25 multiplier, maybe i should drop it to 0

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +121,48 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
             {
-                // Reward sliders based on velocity.
-                sliderBonus = osuLastObj.TravelDistance / osuLastObj.TravelTime;
+                // Take just slider heads into account because we're computing sliderjumps, not slideraim
+                sliderJumpBonus = slider_jump_multiplier * aimStrain * sliderJumpsAdjustingMultiplier;
```
i have no clue what this comment has to do with this line. misplaced?

Givikap120 replied:
this comment explaining why am i using `sliderJumpsAdjustingMultiplier` here
a bit confusing, yes, i will fix that

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -34,19 +36,27 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+            double currVelocity = osuCurrObj.JumpDistance / osuCurrObj.StrainTime;
+
+            // This multiplier exists to prevent slideraim having sliderjumps bonus
+            double sliderJumpsAdjustingMultiplier = 1.0;
```
rename to `sliderJumpNerfFactor` or similar. 

also why not have slider body bonus and slider jump bonus coexist? are there any particular cases you can point to where this doesn't make sense?

Givikap120 replied:
because lost umbrella +100

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +121,48 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+                sliderBonus = osuLastObj.TravelDistance / osuLastObj.TravelTime;
+            }
+
+            aimStrain += sliderBonus * slider_aim_multiplier;
```
if the slider jump bonus and slider aim bonus are not meant to be applied together, then i would suggest throwing them into a max function. this would communicate their purpose much better, and it might render `sliderJumpsAdjustingMultiplier` redundant. 

so i'd be expecting something like:
`aimStrain += Math.max(sliderBonus * slider_aim_multiplier, sliderJumpBonus);`

Givikap120 replied:
they meant to be applied together
sliderjumps account for the fact that aiming sliderhead is harder than normal circle
sliderbody is aiming the body after you hit the circle
sliderJumpsAdjustingMultiplier exist because i don't want sliderbody difficulty to be multiplied by sliderjumps bonus

apollo-dw replied:
so they are meant to be applied together, but also there is a nerf factor that "exists to prevent slideraim having sliderjumps bonus" and you "don't want sliderbody difficulty to be multiplied by sliderjumps bonus". this seems to be a pretty confusing contradiction on a conceptual basis, and i wouldn't expect anyone coming in with a fresh pair of eyes to be able to comprehend what's going on based on the code alone...

Givikap120 replied:
in my opinion it's pretty simple to understand that sliderjump bonus is for aiming to the circle, while slideraim is for aiming the body
aiming the body doesn't gets any harder if previous object was a slider, so i can't just use raw aim strain that includes body difficulty

apollo-dw replied:
i fully understand what slider jump bonus and slider aim bonus is, that doesn't really address my point of there seemingly being contradictions in approach here. if there are no contradictions here and i am mistaken, then the changes here need to be made much clearer

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -112,18 +121,48 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+                // Take just slider heads into account because we're computing sliderjumps, not slideraim
+                sliderJumpBonus = slider_jump_multiplier * aimStrain * sliderJumpsAdjustingMultiplier;
+
+                // Reward more if sliders and circles are alternating (actually it's still lower than several sliders in a row)
+                if (osuLastLastObj.BaseObject is HitCircle)
+                {
+                    double alternatingBonus = slider_jump_multiplier * osuLastObj.JumpDistance / osuLastObj.StrainTime;
+
+                    if (osuLastObj.StrainTime > osuLastLastObj.StrainTime)
+                        alternatingBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuLastObj.StrainTime), 2);
+
+                    sliderJumpBonus += alternatingBonus;
+                }
```
i don't think i'm convinced with this being a factor in difficulty -- could you explain?

Givikap120 replied:
backread a bit

apollo-dw replied:
i can get behind the notion of "slider jumps are more difficult to process" just fine, but i don't agree with alternating between circles and sliders. in my experience, that offers a respite from the increased processing strain more than anything.

> but the cases where there's some notes in the middle of slider jump section or vice versa - it's not becoming any easier to read than just all sliders

the strain algorithm works for this purpose. this shouldn't need any additional bonuses

Givikap120 replied:
try to do what i did:
compare the difficulty of alternating and sliders-only in test diff
alternating case will be considerably underweight compared to sliders-only

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs:
```diff
@@ -56,10 +56,28 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current)
+
+            if (osuCurrObj.BaseObject is Slider slider && osuPrevObj?.BaseObject is Slider)
+            {
+                // Bonus base
+                double sliderStreamBonus = 0.25;
```
unnecessary comment, remove

Givikap120 replied:
ok

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs:
```diff
@@ -56,10 +56,28 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current)
+                if (osuCurrObj.StrainTime > osuPrevObj.StrainTime)
+                    sliderStreamBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuPrevObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuPrevObj.StrainTime), 2);
+
+                // Punish too short sliders to prevent cheesing (cheesing is still possible, but it's very rare)
+                double sliderLength = slider.Velocity * slider.SpanDuration;
+                if (sliderLength < slider.Radius / 2)
+                    sliderStreamBonus *= sliderLength / (slider.Radius / 2);
```
woah -- this is different to the `AimEvaluator` case thanks to the divisions by 2. maybe needs a comment to explain in case this is intentional

Givikap120 replied:
it's intentional cuz sliderstreams usually have shorter sliders than sliderjumps, and i don't want to be as harsh in anti-cheese here

apollo-dw replied:
that doesn't make any sense. so if sliderjump cheese length sliders were used for sliderstreams, they suddenly wouldn't be worthy of receiving cheesing penalty?

Givikap120 replied:
okay, i will change this

apollo-dw commented on osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:
```diff
@@ -31,6 +31,11 @@ public class OsuDifficultyHitObject : DifficultyHitObject
+        /// <summary>
+        /// Normalised distance from the start position of the previous <see cref="OsuDifficultyHitObject"/> to the start position of this <see cref="OsuDifficultyHitObject"/>.
+        /// </summary>
+        public double JumpDistance { get; private set; }
```
this naming is very confusing, because in theory this is even more "lazy" than `LazyJumpDistance` since it's ignoring sliderbodies on sliders

Givikap120 replied:
rename to HeadJumpDistance? CircleJumpDistance?

apollo-dw replied:
i don't like either of those. maybe swapping `JumpDistance` and `LazyJumpDistance` makes more sense. might need more opinions

Givikap120 replied:
i don't like it cuz it will be confusing for people who already know pp
maybe changing LazyJumpDistance to TailJumpDistance?

apollo-dw replied:
that's a fair point. `TailJumpDistance` doesn't quite get there IMO, because it should be general to every object type. something that covers the notion of it being lazy and from the end position would be good. `LazyJumpFromEndDistance` would be ideal for me, but i don't mind long variable names that are descriptive

Givikap120 committed:
Many changes

Refactoring
Changed slow sliders penalty
Made velocity change bonus work correctly

Givikap120 committed:
Merge branch 'master' into slider_jumps

Givikap120 committed:
renamings

Givikap120 committed:
multiple changes

Givikap120 committed:
added sanity check

Givikap120 commented:
Making this Draft because I split-up all changes

Givikap120 committed:
minor refactorings

Givikap120 committed:
improved alternating bonus

Givikap120 committed:
Update AimEvaluator.cs

stanriders commented:
This PR has been split into smaller ones
